services/bots.py

```diff
@@ -58,23 +61,33 @@ def _get_apps_from_weighted_selection(
             ),
             owner.github_app_installations or [],
         )
-    )
+    }
     # We assign weights to the apps based on how long ago they were created.
     # The idea is that there's a greater chance that a change misconfigured the app,
     # So apps recently created are selected less frequently than older apps
+    keys = list(ghapp_installations_filter.keys())
     weights = [
-        min(MAX_GITHUB_APP_SELECTION_WEIGHT, _get_installation_weight(obj))
-        for obj in ghapp_installations_filter
+        min(
+            MAX_GITHUB_APP_SELECTION_WEIGHT,
+            _get_installation_weight(ghapp_installations_filter[key]),
+        )
+        for key in keys
     ]
-    # Random selection of size 3.
-    # If all apps have roughly the same probability of being selected, the array would have different entries.
-    # If 1 app dominates the probability of selection than it would probably be that app repeated 3 times, BUT
-    # from time to time the less frequent one would be selected.
-    apps_to_consider = (
-        random.choices(ghapp_installations_filter, weights=weights, k=3)
-        if len(ghapp_installations_filter) > 0
-        else []
-    )
+    # We pick apps one by one until all apps have been selected
+    # Obviously apps with a higher weight have a higher change of being selected as the main app (1st selection)
+    # But it's important that others are also selected so we can use them as fallbacks
+    apps_to_consider = []
+    if keys:
```

Comment:
Can we remove this `if` check here (i.e. the `while` loop condition should fail if the list is empty)?

Reply:
Sure, I'll refactor this